### ISSUE TITLE
Show MPI connectivity map during MPI_INIT

### ISSUE THREAD

**Issue by ompiteam** [open-mpi/ompi#30]
It has long been discussed, and I swear there was a ticket about this
at some point but I can't find it now.  So I'm filing a new one --
close this as a dupe if someone can find an older one.

---

OMPI currently uses a negative ACK system to indicate if high-speed
networks are not used for MPI communications.  For example, if you
have the openib BTL available but it can't find any active ports in a
given MPI process, it'll display a warning message.

But some users want a ''positive'' acknowledgement of what networks
are being used for MPI communications (this can also help with
regression testing, per a thread on the MTT mailing list).  HP MPI
offers this feature, for example.  It would be nice to have a simple
MCA parameter that will cause MCW rank 0 to output a connectivity map
during MPI_INIT.

Complications:
- In some cases, OMPI doesn't know which networks will be used for
  communications with each MPI process peer; we only know which ones
  we'll try to use when connections are actually established (per
  OMPI's lazy connection model for the OB1 PML).  But I think that
  even outputting this information will be useful.
- Connectivity between MPI processes are likely to be non-uniform.
  E.g., MCW rank 0 may use the sm btl to communicate with some MPI
  processes, but a different btl to communicate with others.  This is
  almost certainly a different view than other processes have.  The
  connectivity information needs to be conveyed on a process-pair
  basis (e.g., a 2D chart).
- Since we have to span multiple PMLs, this may require an addition
  to the PML API.

A first cut could display a simple 2D chart of how OMPI thinks it may
send MPI traffic from each process to each process.  Perhaps something
like (OB1 6 process job, 2 processes on each of 3 hosts):

```
MCW rank 0     1     2     3     4     5
0        self  sm    tcp   tcp   tcp   tcp
1        sm    self  tcp   tcp   tcp   tcp
2        tcp   tcp   self  sm    tcp   tcp
3        tcp   tcp   sm    self  tcp   tcp
4        tcp   tcp   tcp   tcp   self  sm
5        tcp   tcp   tcp   tcp   sm    self
```

Note that the upper and lower triangular portions of the map are the
same, but it's probably more human-readable if both are output.
However, multiple built-in output formats could be useful, such as:
- Human readable, full map (see above)
- Human readable, abbreviated (see below for some ideas on this)
- Machine parsable, full map
- Machine parsable, abbreviated

It may also be worthwhile to investigate a few huersitics to compress
the graph where possible.  Some random ideas in this direction:
- The above example could be represented as:

```
MPI connectivty map, listed by process:
X->X: self
X<->X+1, X in {0,2,4}: sm
other: tcp
```
- Another example:

```
MPI connectivty map, listed by process:
X->X: self
other: tcp
```
- Another example:

```
MPI connectivty map, listed by process:
all: CM PML, MX MTL
```
- Perhaps something could be done with "exceptions" -- e.g., where
  the openib BTL is being used for inter-node connectivity ''except''
  for one node (where IB is malfunctioning, and OMPI fell back to
  TCP) -- this is a common case that users/sysadmins want to detect.

Another useful concept might be to show some information about each
endpoint in the connectivity map.  E.g., show a list of TCP endpoints
on each process, by interface name and/or IP address.  Similar for
other transports.  This kind of information can show when/if
multi-rail scenarios are active, etc.  For example:

```
MCW rank 0     1     2     3     4     5
0        self      sm        tcp:eth0  tcp:eth0  tcp:eth0  tcp:eth0
1        sm        self      tcp:eth0  tcp:eth0  tcp:eth0  tcp:eth0
2        tcp:eth0  tcp:eth0  self      sm        tcp:eth0  tcp:eth0
3        tcp:eth0  tcp:eth0  sm        self      tcp:eth0  tcp:eth0
4        tcp:eth0  tcp:eth0  tcp:eth0  tcp:eth0  self      sm
5        tcp:eth0  tcp:eth0  tcp:eth0  tcp:eth0  sm        self
```

With more information such as interface names, compression of the
output becomes much more important, such as:

```
MPI connectivty map, listed by process:
X->X: self
X<->X+1, X in {0,2,4}: sm
other: tcp:eth0,eth1
```

Note that these ideas can certainly be implemented in stages; there's
no need to do everything at once.

**Comment by jsquyres** [comment]
@gpaulsen @markalle We had a lengthy discussion about this connectivity map yesterday during the [2016 Feb Dallas Open MPI dev meeting](https://github.com/open-mpi/ompi/wiki/Meeting-2016-02), and then a further lengthy conversation about this at dinner last night.  Main points:

* Platform MPI shows several things in their `-prot` output:
  1. See https://gist.github.com/markalle/52fb42c701267c49b450 for an example
  1. Per-host detail of IP address and which MCW rank processes are on that host
  1. Brief NxN chart showing connectivity methods
  1. Processor affinity info
* We agreed that -- at least under a certain size -- users like to see the NxN map.  Printing on a per-server basis seems much more scalable than a per-process basis.
  * The thought is that individual components can report their "short name" (e.g., "SHM", "TCP", "IB", etc.) to be reported in the map.  Maybe put a 3-letter cap on the short name, or somesuch.
* We also like the idea of printing exceptions to that map (i.e., if one server is different than the others)
* We also agree that this functionality must be optional.  It will likely incur additional startup / shutdown cost.
* Platform MPI's affinity info is expressed in terms of physical Linux virtual processor IDs.  We'd want logical core (or hyperthread) IDs for an Open MPI display.
* We'd also like to see, in the per-host detail:
  * A listing of all the network interfaces (perhaps only the interfaces being used...?): eth7, mlx4_0, usnic_1, ...etc.
  * A human name for the network type of each interface: Ethernet, InfiniBand, Omnipath, ...etc.
  * Relevant addressing info for each interface: e.g., for IP-based interfaces, display the IP address/netmask, for IB-based interfaces, the GID+LID.  ...et.
  * Protocol type for the connection: RC, UD, TCP, UDP, ...etc.
* Concerns were brought up that MPI point-to-point connectivity could be different than one-sided and/or collective connectivity.  An idea was floated that perhaps we could have some new infrastructure -- possibly in the MPI base? -- that allows self-reporting of connectivity.
   * E.g., a BTL can make a function call when it makes a connection to a peer identifying:
     * The network interface name
     * Human name for the network type
     * Short name for the connection NxN map
     * Addressing info (as a string)
     * Framework/component name that made the connection
     * Human name for the protocol used
  * In this way, the infrastructure can basically de-duplicate all these reports (e.g., if both a BTL and an OSC report a connection to a common peer), and we'll have an accurate listing of all connections from all types of MPI communications
  * To be clear, BTLs, MTLs, and possibly some of the PMLs will need to report this info.  So will any OSCs that make their own connections, and any COLLs.
  * Further, this infrastructure can be invoked to obtain all the connection information so far.  Two common use cases:
     1. Print this info during MPI_INIT.  In this case, you almost certainly want to do a "preconnect" type of call first, to force all PML-based connections to all peers to be created.  
     1. Print this info during MPI_FINALIZE.  In this case, you do not need "preconnect" functionality in MPI_INIT, but rather only show what connections were *actually* created during the run.
  * In both cases, the info can be gathered -- either to MCW rank 0 or mpirun? -- and prettyprinted nicely.

**Comment by jsquyres** [comment]
IBM is taking on this feature enhancement.

**Comment by jjhursey** [comment]
See PR #2825

**Comment by hppritcha** [comment]
Moving to 3. x as it probably will not get in to 2.1.

**Comment by bwbarrett** [comment]
@jjhursey I'm going to punt this off any milestone, since it looks like it has died out on your side.

**Comment by jjhursey** [comment]
That's fine. I'll add this to the face-to-face meeting to see where we are at again. @markalle maybe we can chat about this again before the meeting sometime.

**Comment by jjhursey** [comment]
Some discussion at the March 2018 Face-to-Face meeting.

Step 1: Display basic table for pt2pt connections (output only)
 * Need to bring in the '-prot' framework
 * Need to add interface to pml/mtl/btl to get a "short name" (e.g., "yalla", "ob1") and a "long name" with information about, for example, interfaces used.
 * For the long names think about aggregation ability.

Step 2: Future
 * Add table for `osc` and maybe `coll` components
 * Another effort to select subsets of components based upon interconnect (`-TCP`, `-UCX`)

@markalle and @jjhursey will investigate a PR for Step 1.

**Comment by jjhursey** [comment]
This was implemented in `master` and `v5.0.x` branches as the `--mca ompi_display_comm` command line option. See the following PRs for more details:
 * Open MPI 'hook' component:
   * https://github.com/open-mpi/ompi/pull/5507
   * https://github.com/open-mpi/ompi/pull/8313
 * PRRTE CLI option
   * https://github.com/openpmix/prrte/pull/711
